lib/Dialect/ESI/ESITranslations.cpp

```diff
@@ -0,0 +1,225 @@
+  void reduceIndent() { currentIndent -= 2; }
+  size_t currentIndent = 0;
+
+  ModuleOp module;
+  llvm::raw_ostream &os;
+  DiagnosticEngine &diag;
+  const Location unknown;
+  size_t errorCount = 0;
+
+  // All the `esi.cosim` input and output types encountered during the IR walk.
+  // This is NOT in a deterministic order!
+  llvm::DenseSet<Type> types;
+};
+} // anonymous namespace
+
+bool ExportCosimSchema::isTypeSupported(Type type) {
+  auto chanPort = type.dyn_cast<ChannelPort>();
+  if (chanPort) {
+    uint64_t innerHash = isTypeSupported(chanPort.getInner());
+    return llvm::hash_combine(chanPort.getTypeID(), innerHash);
+  }
+
+  auto i = type.dyn_cast<IntegerType>();
+  if (i)
+    return i.getWidth() <= 64;
+  return false;
+}
+
+LogicalResult ExportCosimSchema::emitSchemaFor(IntegerType type) {
+  SmallString<16> typeStr;
+  if (type.isSigned())
+    typeStr = "Int";
+  else
+    typeStr = "UInt";
+
+  // Round up.
+  auto w = type.getWidth();
+  if (w == 1)
+    typeStr = "Bool";
+  else if (w <= 8)
+    typeStr += "8";
+  else if (w <= 16)
+    typeStr += "16";
+  else if (w <= 32)
+    typeStr += "32";
+  else if (w <= 64)
+    typeStr += "64";
+  else
+    return diag.emit(unknown, DiagnosticSeverity::Error)
+           << "Capnp does not support ints larger than 64-bit";
+
+  // Since capnp requires messages to be structs, emit a wrapper struct.
+  indent() << "struct ";
+  emitName(type) << " ";
+  emitId(getCapnpTypeID(type)) << " {\n";
+  addIndent();
+
+  // Specify the actual type, followed by the capnp field.
+  indent() << "# Actual type is " << type << ".\n";
+  indent() << "i @0 :" << typeStr << ";\n";
+
+  reduceIndent();
+  indent() << "}\n\n";
+  return success();
+}
+
+LogicalResult ExportCosimSchema::visitEndpoint(CosimEndpoint ep) {
+  ChannelPort inputPort = ep.input().getType().dyn_cast<ChannelPort>();
+  if (!inputPort)
+    return ep.emitOpError("Expected ChannelPort type for input. Got ")
+           << inputPort;
+  if (!isTypeSupported(inputPort))
+    return ep.emitOpError("Type '") << inputPort << "' not supported.";
+  types.insert(inputPort);
+
+  ChannelPort outputPort = ep.output().getType().dyn_cast<ChannelPort>();
+  if (!outputPort)
+    ep.emitOpError("Expected ChannelPort type for output. Got ") << outputPort;
+  if (!isTypeSupported(outputPort))
+    return ep.emitOpError("Type '") << outputPort << "' not supported.";
+  types.insert(outputPort);
+
+  os << "# Endpoint #" << ep.endpointID() << " at " << ep.getLoc() << ":\n";
+  os << "#   Input type: ";
+  emitName(inputPort.getInner()) << " ";
+  emitId(ep.getInputTypeID()) << "\n";
+  os << "#   Output type: ";
+  emitName(outputPort.getInner()) << " ";
+  emitId(ep.getOutputTypeID()) << "\n";
+
+  return success();
+}
+
+LogicalResult ExportCosimSchema::emit() {
+  os << "#########################################################\n"
+     << "## ESI generated schema. For use with CosimDpi.capnp\n"
+     << "#######\n";
```

Comment:
any reason not to fill this line out to match the header?

Reply:
Style. But you're right -- this is non-standard.